File: config/cfg/answerserver/rag/grok.lua
-- grok.lua
-- Answer Server 25.3 RAG → Grok-4 using send_http_request + dkjson
-- Uses ONLY built-in, officially supported libraries

-- dkjson is the standard JSON library in Answer Server (always available)
local json = require("dkjson")

-- === CONFIGURATION ===
local GROK_API_KEY = "YOUR_REAL_GROK_API_KEY_HERE"
local API_URL = "https://api.x.ai/v1/chat/completions"
local MODEL = "grok-4-fast-reasoning"

-- === LOGGING TO YOUR CUSTOM FILE ===
local LOG_FILE = "E:/Knowledge Discovery/AnswerServer/AnswerServer_25.2.0_WINDOWS_X86_64/logs/custom/grok_lua.log"
local function log(msg)
    local f = io.open(LOG_FILE, "a")
    if f then
        f:write(os.date("%Y-%m-%d %H:%M:%S") .. " | " .. tostring(msg) .. "\n")
        f:close()
    end
end

log("=== grok.lua loaded (dkjson version) ===")

-- === GENERATE FUNCTION ===
function generate(prompt, generation_utils)
    log("generate() called – prompt length: " .. #prompt)

    -- Build conversation history from session if available
    local full_prompt = prompt
    if generation_utils and generation_utils.session_data then
        local session = generation_utils:session_data()
        if session and #session > 0 then
            local history = {}
            for _, turn in ipairs(session) do
                table.insert(history, "User: " .. (turn.question or "Unknown"))
                table.insert(history, "Assistant: " .. (turn.answer or "Unknown"))
            end
            full_prompt = table.concat(history, "\n") .. "\n\nCurrent question:\n" .. prompt
            log("Added session history (" .. #session .. " turns)")
        end
    end

    -- Request payload
    local payload = {
        model = MODEL,
        messages = {
            {
                role = "system",
                content = "You are an expert Israel Railways electrical and traction engineer. Answer precisely using only the provided context from technical drawings."
            },
            {
                role = "user",
                content = full_prompt
            }
        },
        temperature = 0.0,
        max_tokens = 10000,
        stream = false
    }

    local payload_json = json.encode(payload)

    log("Sending request to Grok API...")
    local ok, response = pcall(send_http_request, {
        url = API_URL,
        method = "POST",
        headers = {
            ["Content-Type"] = "application/json",
            ["Authorization"] = "Bearer " .. GROK_API_KEY
        },
        content = payload_json,
        section = "ContentSSL"   -- Uses your [ContentSSL] SSLMethod=Negotiate
    })

    if not ok then
        log("HTTP REQUEST FAILED: " .. tostring(response))
        return "Error: Could not connect to AI model. Please try again later."
    end

    if type(response) ~= "string" or response == "" then
        log("Empty or invalid response from API")
        return "Error: Empty response from AI model."
    end

    local success, result = pcall(json.decode, response)
    if not success then
        log("JSON decode failed: " .. response:sub(1, 1000))
        return "Error: Invalid response format from AI model."
    end

    if not result.choices or not result.choices[1] or not result.choices[1].message then
        log("No valid answer in response: " .. response:sub(1, 1000))
        return "Error: AI model returned no answer."
    end

    local answer = result.choices[1].message.content or ""
    log("Answer generated – length: " .. #answer)
    return answer
end

-- === TOKEN COUNTER (required by Answer Server) ===
function get_token_count(text, token_limit)
    local words = {}
    for word in text:gmatch("%S+") do
        table.insert(words, word)
    end

    local count = #words
    local limited_text = count <= token_limit and text or table.concat(words, " ", 1, token_limit)

    return limited_text, count
end

log("grok.lua fully initialized – using dkjson + send_http_request")

File: config/cfg/answerserver/rag/grok.py
import logging
import os
import requests
import json
from logging.handlers import RotatingFileHandler

# --- CONFIGURATION ---
LOG_DIR = r"E:\\Knowledge Discovery\\AnswerServer\\AnswerServer_25.2.0_WINDOWS_X86_64\\logs\\custom"
LOG_FILE = os.path.join(LOG_DIR, "grok_railway_qa.log")

# Ensure log directory exists
os.makedirs(LOG_DIR, exist_ok=True)

# --- SETUP LOCAL FILE + CONSOLE LOGGER ONLY ---
logger = logging.getLogger("grok-railway-qa")
logger.setLevel(logging.DEBUG)  # Capture everything

# Prevent adding handlers multiple times (critical in NiFi/IDOL)
if not logger.handlers:
    # 1. ROTATING FILE HANDLER (main log destination)
    file_handler = RotatingFileHandler(
        LOG_FILE,
        maxBytes=5_000_000,      # 5 MB per file
        backupCount=10,          # Keep 10 old logs
        encoding='utf-8'
    )
    file_formatter = logging.Formatter(
        '%(asctime)s | %(levelname)-8s | %(filename)s:%(lineno)d | %(message)s'
    )
    file_handler.setFormatter(file_formatter)
    file_handler.setLevel(logging.DEBUG)
    logger.addHandler(file_handler)

    # 2. CONSOLE HANDLER (for immediate visibility when running locally)
    console_handler = logging.StreamHandler()
    console_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
    console_handler.setFormatter(console_formatter)
    console_handler.setLevel(logging.INFO)
    logger.addHandler(console_handler)

    logger.info("LOCAL FILE + CONSOLE LOGGING INITIALIZED")
    logger.info(f"Log file: {LOG_FILE}")
    logger.debug("Debug mode active – all details will be written to log file")

# --- CONSTANTS ---
API_KEY = "YOUR_REAL_GROK_API_KEY_HERE"

# --- MAIN GENERATE FUNCTION ---
def generate(prompt: str) -> str:
    logger.debug(f"generate() called – prompt length: {len(prompt)} characters")
    logger.debug(f"Prompt preview: {prompt[:500]}{'...' if len(prompt) > 500 else ''}")

    try:
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {API_KEY}"
        }

        data = {
            "messages": [
                {"role": "system", "content": "You are an expert railway electrical engineer. Answer precisely using only the provided drawing context."},
                {"role": "user", "content": prompt}
            ],
            "model": "grok-4-fast-reasoning",
            "stream": False,
            "temperature": 0.0,
            "max_tokens": 10000
        }

        logger.info("Sending request to xAI Grok API...")
        response = requests.post(
            "https://api.x.ai/v1/chat/completions",
            headers=headers,
            json=data,
            timeout=3600
        )

        logger.info(f"API Response: {response.status_code} {response.reason}")
        logger.debug(f"Response headers: {dict(response.headers)}")

        response.raise_for_status()
        result = response.json()

        if "choices" not in result or not result["choices"]:
            logger.error("API returned no 'choices'")
            raise ValueError("Invalid response from Grok API")

        content = result["choices"][0]["message"]["content"]
        logger.info(f"Answer received – {len(content)} characters")
        logger.debug(f"Answer preview: {content[:500]}")

        return content

    except requests.exceptions.RequestException as e:
        logger.error(f"HTTP/network error: {e}", exc_info=True)
        raise
    except Exception as e:
        logger.error(f"Unexpected error in generate(): {e}", exc_info=True)
        raise

# --- TOKEN LIMIT HELPER ---
def get_token_count(text: str, token_limit: int):
    logger.debug(f"get_token_count called – original text length: {len(text)}")
    words = text.split()
    limited_text = ' '.join(words[:token_limit])
    actual_count = len(words)
    logger.debug(f"Text limited to {token_limit} words → result length: {len(limited_text)}, total words were {actual_count}")
    return limited_text, actual_count

# --- Test on load (optional) ---
if __name__ == "__main__":
    logger.info("Script loaded successfully – logging to local file only")
    logger.info(f"Log location: {LOG_FILE}")